app/src/main/java/nu/yona/app/security/EncryptionKeyGenerator.java

```diff
@@ -0,0 +1,158 @@
+			{
+				return getExistingSecurityKeyFromAlias(keyStore);
+			}
+			return getFreshSecurityKey(keyStore);
+		}
+		catch (KeyStoreException e)
+		{
+			AppUtils.reportException(EncryptionKeyGenerator.class, e, Thread.currentThread(), null, false);
+			throw new YonaRuntimeException(e);
+		}
+	}
+
+	@TargetApi(Build.VERSION_CODES.M)
+	static SecurityKey getExistingSecurityKeyFromAlias(KeyStore keyStore)
+	{
+		try
+		{
+			KeyStore.SecretKeyEntry entry =
+					(KeyStore.SecretKeyEntry) keyStore.getEntry(KEY_ALIAS, null);
+			return new SecurityKey(entry.getSecretKey());
+		}
+		catch (KeyStoreException | NoSuchAlgorithmException | UnrecoverableEntryException e)
+		{
+			AppUtils.reportException(EncryptionKeyGenerator.class, e, Thread.currentThread(), null, false);
+			throw new YonaRuntimeException(e);
+		}
+	}
+
+
+	@TargetApi(Build.VERSION_CODES.M)
+	static SecurityKey getFreshSecurityKey(KeyStore keyStore)
+	{
+		try
+		{
+			KeyGenerator keyGenerator =
+					KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, ANDROID_KEY_STORE);
+			keyGenerator.init(new KeyGenParameterSpec.Builder(KEY_ALIAS,
+					KeyProperties.PURPOSE_ENCRYPT | KeyProperties.PURPOSE_DECRYPT).setBlockModes(
+					KeyProperties.BLOCK_MODE_GCM)
+					.setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
+					.setRandomizedEncryptionRequired(false)
+					.build());
+			return new SecurityKey(keyGenerator.generateKey());
+		}
+		catch (NoSuchProviderException | NoSuchAlgorithmException | InvalidAlgorithmParameterException e)
+		{
+			AppUtils.reportException(EncryptionKeyGenerator.class, e, Thread.currentThread(), null, false);
+			throw new YonaRuntimeException(e);
+		}
+	}
+
+	@TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
+	static SecurityKey generateSecretKeyPreM(Context context, KeyStore keyStore)
+	{
+		try
+		{
+			if (keyStore.containsAlias(KEY_ALIAS))
+			{
+				return getExistingSecretKeyPreM(keyStore);
+			}
+			return getFreshSecretKeyPreM(context);
+		}
+		catch (KeyStoreException e)
+		{
+			AppUtils.reportException(EncryptionKeyGenerator.class, e, Thread.currentThread(), null, false);
+			throw new YonaRuntimeException(e);
+		}
+	}
+
+	@TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
+	static SecurityKey getExistingSecretKeyPreM(KeyStore keyStore)
+	{
+		try
+		{
+			KeyStore.PrivateKeyEntry entry =
+					(KeyStore.PrivateKeyEntry) keyStore.getEntry(KEY_ALIAS, null);
+			return new SecurityKey(
+					new KeyPair(entry.getCertificate().getPublicKey(), entry.getPrivateKey()));
+		}
+		catch (KeyStoreException | NoSuchAlgorithmException | UnrecoverableEntryException e)
+		{
+			AppUtils.reportException(EncryptionKeyGenerator.class, e, Thread.currentThread(), null, false);
+			throw new YonaRuntimeException(e);
+		}
+	}
+
+	@TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
+	static SecurityKey getFreshSecretKeyPreM(Context context)
+	{
+		try
+		{
+			Calendar start = Calendar.getInstance();
+			Calendar end = Calendar.getInstance();
+			//1 Year validity
```

Comment:
Why only one year? We might be supporting pre-M devices for more than a year, right? Can't you make it 5?

Reply:
Done.